source/Halibut/Transport/TcpClientExtensions.cs

```diff
@@ -1,46 +1,68 @@
 using System;
 using System.Net.Sockets;
-using Halibut.Diagnostics;
 using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Halibut.Util.AsyncEx;
 
 namespace Halibut.Transport
 {
     public static class TcpClientExtensions
     {
         public static void ConnectWithTimeout(this TcpClient client, Uri remoteUri, TimeSpan timeout)
         {
-            client.ConnectWithTimeout(remoteUri.Host, remoteUri.Port, timeout);
+            ConnectWithTimeout(client, remoteUri.Host, remoteUri.Port, timeout, CancellationToken.None);
+        }
+
+        public static void ConnectWithTimeout(this TcpClient client, Uri remoteUri, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            ConnectWithTimeout(client, remoteUri.Host, remoteUri.Port, timeout, cancellationToken);
         }
 
         public static void ConnectWithTimeout(this TcpClient client, string host, int port, TimeSpan timeout)
         {
-            var connectResult = false;
+            ConnectWithTimeout(client, host, port, timeout, CancellationToken.None);
+        }
+
+        public static void ConnectWithTimeout(this TcpClient client, string host, int port, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            Connect(client, host, port, timeout, cancellationToken).GetAwaiter().GetResult();
+        }
+
+        static async Task Connect(TcpClient client, string host, int port, TimeSpan timeout, CancellationToken cancellationToken)
+        {
             try
             {
-                connectResult = client.ConnectAsync(host, port).Wait(timeout);
+                var task = client.ConnectAsync(host, port);
+                await task.TimeoutAfter(timeout, cancellationToken);
```

Comment:
Most of the code in this file comes from This comes from the related PR https://github.com/OctopusDeploy/Halibut/pull/109.

The PR is all about handing round the cancellation token so we can get a real one to this line.